src/presentation.ts

```diff
@@ -14,178 +22,223 @@ import {
+  let claims = deserializeNestedProvableValue(request.claims);
+
+  switch (request.type) {
+    case 'no-context':
+      return PresentationRequest.noContext(spec, claims);
+    case 'zk-app': {
+      const inputContext: any = deserializeInputContext(request.inputContext);
+      return ZkAppRequest({ spec, claims, inputContext });
+    }
+    case 'https': {
+      const inputContext: any = deserializeInputContext(request.inputContext);
+      return HttpsRequest({ spec, claims, inputContext });
+    }
+    default:
+      throw Error(`Invalid presentation request type: ${request.type}`);
+  }
+}
+
 type Presentation<Output, Inputs extends Record<string, Input>> = {
   version: 'v0';
   claims: Claims<Inputs>;
   outputClaim: Output;
+  clientNonce: Field;
   proof: Proof<PublicInputs<Inputs>, Output>;
 };
 
-type Output<R> = R extends PresentationRequest<infer O> ? O : never;
-type Inputs<R> = R extends PresentationRequest<any, infer I> ? I : never;
+type Output<R> = R extends PresentationRequest<any, infer O> ? O : never;
+type Inputs<R> = R extends PresentationRequest<any, any, infer I> ? I : never;
+type WalletContext<R> = R extends PresentationRequest<
+  any,
+  any,
+  any,
+  any,
+  infer W
+>
+  ? W
+  : never;
 
 const Presentation = {
   async compile<R extends PresentationRequest>(
     request: R
-  ): Promise<R & { program: Program<Output<R>, Inputs<R>> }> {
+  ): Promise<
+    Omit<R, 'program'> & {
+      program: Program<Output<R>, Inputs<R>>;
+      verificationKey: VerificationKey;
+    }
+  > {
     let program: Program<Output<R>, Inputs<R>> = (request as any).program ??
     createProgram(request.spec);
-    await program.compile();
-    return { ...request, program };
+    let verificationKey = await program.compile();
+    return { ...request, program, verificationKey };
   },
 
   create: createPresentation,
 };
 
-async function createPresentation<Output, Inputs extends Record<string, Input>>(
+async function createPresentation<R extends PresentationRequest>(
   ownerKey: PrivateKey,
   {
     request,
-    walletContext,
+    context: walletContext,
     credentials,
   }: {
-    request: PresentationRequest<Output, Inputs>;
-    walletContext?: ZkAppWalletContext | HttpsWalletContext;
+    request: R;
+    context: WalletContext<R>;
     credentials: (StoredCredential & { key?: string })[];
   }
-): Promise<Presentation<Output, Inputs>> {
-  let context = request.deriveContext(walletContext);
-  let { program } = await Presentation.compile(request);
+): Promise<Presentation<Output<R>, Inputs<R>>> {
+  // compile the program
+  let { program, verificationKey } = await Presentation.compile(request);
 
+  // generate random client nonce
+  let clientNonce = Field.random();
+
+  // derive context
+  let context = request.deriveContext(request.inputContext, walletContext, {
+    clientNonce,
+    vkHash: verificationKey.hash,
+    claims: hashClaims(request.claims),
+  });
```

Comment:
vk hash and claims are derived client-side from the request

client nonce is generated automatically